File: lib/chat/sessions/store.ex
# Назначение файла: транзакционные переходы устойчивого состояния чат-сессии.
defmodule Chat.Sessions.Store do
  import Ecto.Query

  alias Chat.Repo
  alias Chat.Sessions.ChatSession

  @grace_seconds 60
  @hidden_grace_seconds 30 * 60
  @heartbeat_timeout_seconds 180

  def hidden_grace_seconds, do: @hidden_grace_seconds

  def create(attrs) do
    session_id = Map.fetch!(attrs, :id)
    Repo.insert(ChatSession.create_changeset(%ChatSession{id: session_id}, attrs))
  end

  def end_active_for_identity(room_id, identity_key, now \\ DateTime.utc_now()) do
    now = truncate(now)

    from(session in ChatSession,
      where:
        session.room_id == ^room_id and session.identity_key == ^identity_key and
          session.status in ["active", "reconnecting"]
    )
    |> Repo.update_all(
      set: [status: "ended", ended_at: now, reconnect_deadline_at: nil, updated_at: now],
      inc: [generation: 1]
    )
  end

  def restore(session_id, identity_key, resume_secret, now \\ DateTime.utc_now(), opts \\ []) do
    now = truncate(now)

    Repo.transaction(fn ->
      with {:ok, id} <- Ecto.UUID.cast(session_id),
           %ChatSession{} = session <- Repo.get(ChatSession, id),
           true <- session.identity_key == identity_key,
           true <- valid_secret?(session, resume_secret),
           true <- session.room_id == Keyword.get(opts, :room_id, session.room_id),
           true <- session.nickname == Keyword.get(opts, :nickname, session.nickname),
           false <- session.status == "ended",
           false <- expired?(session, now) do
        result =
          ChatSession
          |> where(
            [current],
            current.id == ^session.id and current.generation == ^session.generation and
              current.status in ["active", "reconnecting"] and
              (is_nil(current.reconnect_deadline_at) or current.reconnect_deadline_at > ^now)
          )
          |> Repo.update_all(
            set: [
              status: "active",
              last_seen_at: now,
              reconnect_deadline_at: nil,
              updated_at: now
            ],
            inc: [generation: 1]
          )

        case result do
          {1, _} -> Repo.get!(ChatSession, session.id)
          {0, _} -> Repo.rollback(:session_ended)
        end
      else
        nil -> Repo.rollback(:not_found)
        false -> Repo.rollback(:invalid_session)
        true -> Repo.rollback(:session_ended)
        _ -> Repo.rollback(:invalid_session)
      end
    end)
  end

  def reconnect(session_id, identity_key, generation, now \\ DateTime.utc_now()) do
    now = truncate(now)

    query =
      from session in ChatSession,
        where:
          session.id == ^session_id and session.identity_key == ^identity_key and
            session.generation == ^generation and session.status == "active"

    case Repo.one(query) do
      nil ->
        :stale

      session ->
        deadline = DateTime.add(now, grace_seconds(session), :second)

        case Repo.update_all(query,
               set: [status: "reconnecting", reconnect_deadline_at: deadline, updated_at: now]
             ) do
          {1, _} -> {:ok, deadline}
          {0, _} -> :stale
        end
    end
  end

  def activate(session_id, identity_key, generation, now \\ DateTime.utc_now()) do
    now = truncate(now)
    cutoff = DateTime.add(now, -(@heartbeat_timeout_seconds + @grace_seconds), :second)

    hidden_cutoff =
      DateTime.add(now, -(@heartbeat_timeout_seconds + @hidden_grace_seconds), :second)

    from(session in ChatSession,
      where:
        session.id == ^session_id and session.identity_key == ^identity_key and
          session.generation == ^generation and session.status in ["active", "reconnecting"] and
          ((session.status == "active" and
              ((session.last_visibility == "hidden" and session.last_seen_at > ^hidden_cutoff) or
                 (session.last_visibility != "hidden" and session.last_seen_at > ^cutoff))) or
             (session.status == "reconnecting" and session.reconnect_deadline_at > ^now)),
      select: session.generation
    )
    |> Repo.update_all(
      set: [
        status: "active",
        last_seen_at: now,
        reconnect_deadline_at: nil,
        updated_at: now
      ],
      inc: [generation: 1]
    )
    |> case do
      {1, [epoch]} -> {:ok, epoch}
      {0, _} -> {:error, :stale_connection}
    end
  end

  def end_session(session_id, identity_key, generation, now \\ DateTime.utc_now()) do
    now = truncate(now)

    query =
      from session in ChatSession,
        where:
          session.id == ^session_id and session.identity_key == ^identity_key and
            session.generation == ^generation and session.status in ["active", "reconnecting"]

    case Repo.update_all(query,
           set: [status: "ended", ended_at: now, reconnect_deadline_at: nil, updated_at: now],
           inc: [generation: 1]
         ) do
      {1, _} -> :ended
      {0, _} -> :stale
    end
  end

  def expired(now \\ DateTime.utc_now()) do
    now = truncate(now)

    from(session in ChatSession,
      where: session.status == "reconnecting" and session.reconnect_deadline_at <= ^now
    )
    |> Repo.all()
  end

  def expire(session, now \\ DateTime.utc_now()) do
    now = truncate(now)

    query =
      from current in ChatSession,
        where:
          current.id == ^session.id and current.generation == ^session.generation and
            current.status == "reconnecting" and
            current.reconnect_deadline_at <= ^now

    case Repo.update_all(query,
           set: [status: "ended", ended_at: now, reconnect_deadline_at: nil, updated_at: now],
           inc: [generation: 1]
         ) do
      {1, _} -> :ended
      {0, _} -> :stale
    end
  end

  defp valid_secret?(%ChatSession{resume_secret_hash: hash}, secret)
       when is_binary(secret) and byte_size(secret) >= 32 do
    Plug.Crypto.secure_compare(hash, hash_secret(secret))
  end

  defp valid_secret?(_session, _secret), do: false

  def hash_secret(secret) when is_binary(secret),
    do: :crypto.hash(:sha256, secret) |> Base.encode16(case: :lower)

  def current?(id, identity, generation) do
    now = DateTime.utc_now() |> truncate()
    cutoff = DateTime.add(now, -(@heartbeat_timeout_seconds + @grace_seconds), :second)

    hidden_cutoff =
      DateTime.add(now, -(@heartbeat_timeout_seconds + @hidden_grace_seconds), :second)

    Repo.exists?(
      from session in ChatSession,
        where:
          session.id == ^id and session.identity_key == ^identity and
            session.generation == ^generation and session.status == "active" and
            ((session.last_visibility == "hidden" and session.last_seen_at > ^hidden_cutoff) or
               (session.last_visibility != "hidden" and session.last_seen_at > ^cutoff))
    )
  end

  def register_identity(session, user, secret) do
    identity = Chat.Visits.user_identity_key(user)
    now = DateTime.utc_now() |> truncate()

    query =
      from current in ChatSession,
        where:
          current.id == ^session.session_id and current.identity_key == ^session.identity_key and
            current.generation == ^session.connection_epoch and current.status == "active",
        select: current

    case Repo.update_all(query,
           set: [
             identity_key: identity,
             resume_secret_hash: hash_secret(secret),
             last_seen_at: now,
             updated_at: now
           ],
           inc: [generation: 1]
         ) do
      {1, [stored]} -> {:ok, stored}
      {0, _} -> {:error, :stale_connection}
    end
  end

  def touch(id, identity, generation),
    do: touch(id, identity, generation, "unknown", DateTime.utc_now())

  def touch(id, identity, generation, %DateTime{} = now),
    do: touch(id, identity, generation, "unknown", now)

  def touch(id, identity, generation, visibility),
    do: touch(id, identity, generation, visibility, DateTime.utc_now())

  def touch(id, identity, generation, visibility, now) do
    now = truncate(now)
    cutoff = DateTime.add(now, -(@heartbeat_timeout_seconds + @grace_seconds), :second)

    hidden_cutoff =
      DateTime.add(now, -(@heartbeat_timeout_seconds + @hidden_grace_seconds), :second)

    visibility = normalize_visibility(visibility)

    query =
      from session in ChatSession,
        where:
          session.id == ^id and session.identity_key == ^identity and
            session.generation == ^generation and session.status == "active" and
            ((session.last_visibility == "hidden" and session.last_seen_at > ^hidden_cutoff) or
               (session.last_visibility != "hidden" and session.last_seen_at > ^cutoff))

    case Repo.update_all(query,
           set: [last_seen_at: now, last_visibility: visibility, updated_at: now]
         ) do
      {1, _} -> :ok
      {0, _} -> :stale
    end
  end

  def mark_stale(now \\ DateTime.utc_now()) do
    now = truncate(now)
    cutoff = DateTime.add(now, -@heartbeat_timeout_seconds, :second)

    query =
      from session in ChatSession,
        where: session.status == "active" and session.last_seen_at <= ^cutoff,
        select: session

    query
    |> Repo.all()
    |> Enum.flat_map(fn session ->
      deadline =
        DateTime.add(
          session.last_seen_at,
          @heartbeat_timeout_seconds + grace_seconds(session),
          :second
        )

      update =
        from current in ChatSession,
          where:
            current.id == ^session.id and current.generation == ^session.generation and
              current.status == "active" and current.last_seen_at <= ^cutoff

      case Repo.update_all(update,
             set: [status: "reconnecting", reconnect_deadline_at: deadline, updated_at: now]
           ) do
        {1, _} -> [session]
        {0, _} -> []
      end
    end)
  end

  def live(room_id) do
    Repo.all(
      from session in ChatSession,
        where: session.room_id == ^room_id and session.status in ["active", "reconnecting"]
    )
  end

  def pending(room_id), do: Enum.filter(live(room_id), &(&1.status == "reconnecting"))

  defp expired?(%ChatSession{reconnect_deadline_at: nil, last_seen_at: seen} = session, now),
    do: DateTime.diff(now, seen) >= @heartbeat_timeout_seconds + grace_seconds(session)

  defp expired?(%ChatSession{reconnect_deadline_at: deadline}, now),
    do: DateTime.compare(deadline, now) != :gt

  defp truncate(datetime), do: DateTime.truncate(datetime, :second)

  defp grace_seconds(%ChatSession{last_visibility: "hidden"}), do: @hidden_grace_seconds
  defp grace_seconds(_session), do: @grace_seconds

  defp normalize_visibility(visibility) when visibility in ["visible", "hidden"], do: visibility
  defp normalize_visibility(_visibility), do: "unknown"
end

File: apps/api/internal/chatsessions/adapters/postgres/store.go
// Package postgres persists fenced chat-session transitions in the shared database.
package postgres

import (
	"context"
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"fmt"
	"strconv"
	"time"

	"chat/api/internal/chatsessions/domain"

	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgconn"
)

var ErrInvalidSession = domain.ErrInvalidSession

type Database interface {
	Begin(context.Context) (pgx.Tx, error)
	QueryRow(context.Context, string, ...any) pgx.Row
	Query(context.Context, string, ...any) (pgx.Rows, error)
	Exec(context.Context, string, ...any) (pgconn.CommandTag, error)
}
type Store struct{ pool Database }

func NewStore(pool Database) Store { return Store{pool: pool} }

func (s Store) Start(ctx context.Context, session domain.Session, resumeSecret string, userID *int64) (domain.Session, error) {
	tx, err := s.pool.Begin(ctx)
	if err != nil {
		return domain.Session{}, fmt.Errorf("begin chat session start: %w", err)
	}
	defer func() { _ = tx.Rollback(ctx) }()
	const visitQuery = `INSERT INTO visits (nickname, identity_key, session_id, user_id, entered_at, inserted_at, updated_at)
	VALUES ($1, $2, $3, $4, (NOW() AT TIME ZONE 'UTC'), (NOW() AT TIME ZONE 'UTC'), (NOW() AT TIME ZONE 'UTC')) RETURNING id`
	if err := tx.QueryRow(ctx, visitQuery, session.Nickname, session.IdentityKey, session.ID, userID).Scan(&session.VisitID); err != nil {
		return domain.Session{}, startError("create visit", err)
	}
	const sessionQuery = `INSERT INTO chat_sessions (id, room_id, identity_key, nickname, resume_secret_hash, status, last_seen_at, generation, visit_id, inserted_at, updated_at)
	VALUES ($1, $2, $3, $4, $5, 'active', (NOW() AT TIME ZONE 'UTC'), 0, $6, (NOW() AT TIME ZONE 'UTC'), (NOW() AT TIME ZONE 'UTC')) RETURNING generation`
	if err := tx.QueryRow(ctx, sessionQuery, session.ID, session.RoomID, session.IdentityKey, session.Nickname, hashSecret(resumeSecret), session.VisitID).Scan(&session.Generation); err != nil {
		return domain.Session{}, startError("create chat session", err)
	}
	if _, err := tx.Exec(ctx, `INSERT INTO room_messages (room_id,kind,author,body,theme_id,appearance,reactions,font_id,font_style,sent_at,inserted_at,updated_at) VALUES ($1,'system','system',$2,'vertigo','{}','{}','theme','normal',(NOW() AT TIME ZONE 'UTC'),(NOW() AT TIME ZONE 'UTC'),(NOW() AT TIME ZONE 'UTC'))`, session.RoomID, "в чат заходит "+session.Nickname); err != nil {
		return domain.Session{}, fmt.Errorf("persist entrance: %w", err)
	}
	if err := tx.Commit(ctx); err != nil {
		return domain.Session{}, fmt.Errorf("commit chat session start: %w", err)
	}
	session.Status = domain.StatusActive
	return session, nil
}

func (s Store) Restore(ctx context.Context, sessionID, identityKey, resumeSecret string, now time.Time, activeCutoff time.Time, hiddenCutoff time.Time) (domain.Session, error) {
	const query = `UPDATE chat_sessions SET status = 'active', last_seen_at = $4, reconnect_deadline_at = NULL, generation = generation + 1, updated_at = $4
	WHERE id = $1 AND identity_key = $2 AND resume_secret_hash = $3 AND status IN ('active', 'reconnecting')
	AND ((status = 'reconnecting' AND reconnect_deadline_at > $4) OR
	     (status = 'active' AND ((last_visibility = 'hidden' AND last_seen_at > $6) OR
	                              (COALESCE(last_visibility, 'unknown') != 'hidden' AND last_seen_at > $5))))
	RETURNING id, room_id, identity_key, nickname, status, generation, visit_id, reconnect_deadline_at`
	var session domain.Session
	if err := s.pool.QueryRow(ctx, query, sessionID, identityKey, hashSecret(resumeSecret), now.UTC(), activeCutoff.UTC(), hiddenCutoff.UTC()).Scan(
		&session.ID, &session.RoomID, &session.IdentityKey, &session.Nickname, &session.Status, &session.Generation, &session.VisitID, &session.ReconnectDeadline,
	); err != nil {
		if errors.Is(err, pgx.ErrNoRows) {
			return domain.Session{}, ErrInvalidSession
		}
		return domain.Session{}, fmt.Errorf("restore chat session: %w", err)
	}
	return session, nil
}

func (s Store) Reconnect(ctx context.Context, sessionID, identityKey string, generation int, now time.Time, grace time.Duration, hiddenGrace time.Duration) (domain.Session, error) {
	const query = `UPDATE chat_sessions SET status = 'reconnecting', reconnect_deadline_at = $4::timestamp +
	CASE WHEN last_visibility = 'hidden' THEN ($6 * interval '1 second') ELSE ($5 * interval '1 second') END, updated_at = $4
	WHERE id = $1 AND identity_key = $2 AND generation = $3 AND status = 'active'
	RETURNING id, room_id, identity_key, nickname, status, generation, visit_id, reconnect_deadline_at`
	return s.transition(ctx, "start reconnect", query, sessionID, identityKey, generation, now.UTC(), int64(grace/time.Second), int64(hiddenGrace/time.Second))
}

func (s Store) Activate(ctx context.Context, sessionID, identityKey string, generation int, now time.Time, activeCutoff time.Time, hiddenCutoff time.Time) (domain.Session, error) {
	const query = `UPDATE chat_sessions SET status = 'active', last_seen_at = $4, reconnect_deadline_at = NULL, generation = generation + 1, updated_at = $4
	WHERE id = $1 AND identity_key = $2 AND generation = $3 AND status IN ('active', 'reconnecting')
	AND ((status = 'reconnecting' AND reconnect_deadline_at > $4) OR
	     (status = 'active' AND ((last_visibility = 'hidden' AND last_seen_at > $6) OR
	                              (COALESCE(last_visibility, 'unknown') != 'hidden' AND last_seen_at > $5))))
	RETURNING id, room_id, identity_key, nickname, status, generation, visit_id, reconnect_deadline_at`
	return s.transition(ctx, "activate chat session", query, sessionID, identityKey, generation, now.UTC(), activeCutoff.UTC(), hiddenCutoff.UTC())
}

func (s Store) Touch(ctx context.Context, sessionID, identityKey string, generation int, visibility string, now time.Time, activeCutoff time.Time, hiddenCutoff time.Time) error {
	if visibility != "visible" && visibility != "hidden" {
		visibility = "unknown"
	}
	const query = `UPDATE chat_sessions SET last_seen_at = $5, last_visibility = $4, updated_at = $5
	WHERE id = $1 AND identity_key = $2 AND generation = $3 AND status = 'active'
	AND ((last_visibility = 'hidden' AND last_seen_at > $7) OR
	     (COALESCE(last_visibility, 'unknown') != 'hidden' AND last_seen_at > $6))
	RETURNING visit_id`
	tx, err := s.pool.Begin(ctx)
	if err != nil {
		return fmt.Errorf("begin chat session touch: %w", err)
	}
	defer func() { _ = tx.Rollback(ctx) }()
	var visitID int64
	if err := tx.QueryRow(ctx, query, sessionID, identityKey, generation, visibility, now.UTC(), activeCutoff.UTC(), hiddenCutoff.UTC()).Scan(&visitID); err != nil {
		return transitionError("touch chat session", err)
	}
	if _, err := tx.Exec(ctx, `UPDATE visits SET updated_at = $2 WHERE id = $1 AND left_at IS NULL`, visitID, now.UTC()); err != nil {
		return fmt.Errorf("touch visit: %w", err)
	}
	if err := tx.Commit(ctx); err != nil {
		return fmt.Errorf("commit chat session touch: %w", err)
	}
	return nil
}

func (s Store) RegisterIdentity(ctx context.Context, sessionID, identityKey string, generation int, userID int64, resumeSecret string, now time.Time) (domain.Session, error) {
	tx, err := s.pool.Begin(ctx)
	if err != nil {
		return domain.Session{}, fmt.Errorf("begin chat identity registration: %w", err)
	}
	defer func() { _ = tx.Rollback(ctx) }()
	registeredIdentity := "user:" + strconv.FormatInt(userID, 10)
	const sessionQuery = `UPDATE chat_sessions SET identity_key = $4, resume_secret_hash = $5, last_seen_at = $6, updated_at = $6, generation = generation + 1
	WHERE id = $1 AND identity_key = $2 AND generation = $3 AND status = 'active'
	RETURNING id, room_id, identity_key, nickname, status, generation, visit_id, reconnect_deadline_at`
	var session domain.Session
	if err := tx.QueryRow(ctx, sessionQuery, sessionID, identityKey, generation, registeredIdentity, hashSecret(resumeSecret), now.UTC()).Scan(
		&session.ID, &session.RoomID, &session.IdentityKey, &session.Nickname, &session.Status, &session.Generation, &session.VisitID, &session.ReconnectDeadline,
	); err != nil {
		return domain.Session{}, transitionError("register chat identity", err)
	}
	const visitQuery = `UPDATE visits SET user_id = $2, identity_key = $3, updated_at = $4 WHERE id = $1 AND left_at IS NULL`
	if tag, err := tx.Exec(ctx, visitQuery, session.VisitID, userID, registeredIdentity, now.UTC()); err != nil {
		return domain.Session{}, fmt.Errorf("register visit identity: %w", err)
	} else if tag.RowsAffected() != 1 {
		return domain.Session{}, ErrInvalidSession
	}
	if err := tx.Commit(ctx); err != nil {
		return domain.Session{}, fmt.Errorf("commit chat identity registration: %w", err)
	}
	return session, nil
}

func (s Store) MarkStale(ctx context.Context, now time.Time, heartbeat time.Duration, grace time.Duration, hiddenGrace time.Duration) ([]domain.Session, error) {
	const query = `UPDATE chat_sessions SET status = 'reconnecting', reconnect_deadline_at = last_seen_at +
	CASE WHEN last_visibility = 'hidden' THEN (($4 + $2) * interval '1 second') ELSE (($3 + $2) * interval '1 second') END, updated_at = $1
	WHERE status = 'active' AND last_seen_at <= $1::timestamp - ($2 * interval '1 second')
	RETURNING id, room_id, identity_key, nickname, status, generation, visit_id, reconnect_deadline_at`
	return s.sessions(ctx, "mark stale chat sessions", query, now.UTC(), int64(heartbeat/time.Second), int64(grace/time.Second), int64(hiddenGrace/time.Second))
}

func (s Store) Expired(ctx context.Context, now time.Time) ([]domain.Session, error) {
	const query = `SELECT id, room_id, identity_key, nickname, status, generation, visit_id, reconnect_deadline_at
	FROM chat_sessions WHERE status = 'reconnecting' AND reconnect_deadline_at <= $1`
	return s.sessions(ctx, "list expired chat sessions", query, now.UTC())
}

func (s Store) sessions(ctx context.Context, action string, query string, args ...any) ([]domain.Session, error) {
	rows, err := s.pool.Query(ctx, query, args...)
	if err != nil {
		return nil, fmt.Errorf("%s: %w", action, err)
	}
	defer rows.Close()
	var sessions []domain.Session
	for rows.Next() {
		var session domain.Session
		if err := rows.Scan(&session.ID, &session.RoomID, &session.IdentityKey, &session.Nickname, &session.Status, &session.Generation, &session.VisitID, &session.ReconnectDeadline); err != nil {
			return nil, fmt.Errorf("scan %s: %w", action, err)
		}
		sessions = append(sessions, session)
	}
	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("iterate %s: %w", action, err)
	}
	return sessions, nil
}

func (s Store) End(ctx context.Context, sessionID, identityKey string, generation int, now time.Time) (domain.Session, error) {
	tx, err := s.pool.Begin(ctx)
	if err != nil {
		return domain.Session{}, fmt.Errorf("begin chat session end: %w", err)
	}
	defer func() { _ = tx.Rollback(ctx) }()
	const sessionQuery = `UPDATE chat_sessions SET status = 'ended', ended_at = $4, reconnect_deadline_at = NULL, generation = generation + 1, updated_at = $4
	WHERE id = $1 AND identity_key = $2 AND generation = $3 AND status IN ('active', 'reconnecting')
	RETURNING id, room_id, identity_key, nickname, status, generation, visit_id, reconnect_deadline_at`
	var session domain.Session
	if err := tx.QueryRow(ctx, sessionQuery, sessionID, identityKey, generation, now.UTC()).Scan(
		&session.ID, &session.RoomID, &session.IdentityKey, &session.Nickname, &session.Status, &session.Generation, &session.VisitID, &session.ReconnectDeadline,
	); err != nil {
		if errors.Is(err, pgx.ErrNoRows) {
			const endedQuery = `SELECT id, room_id, identity_key, nickname, status, generation, visit_id, reconnect_deadline_at
			FROM chat_sessions WHERE id = $1 AND identity_key = $2 AND status = 'ended'`
			if endedErr := tx.QueryRow(ctx, endedQuery, sessionID, identityKey).Scan(
				&session.ID, &session.RoomID, &session.IdentityKey, &session.Nickname, &session.Status, &session.Generation, &session.VisitID, &session.ReconnectDeadline,
			); endedErr == nil {
				if commitErr := tx.Commit(ctx); commitErr != nil {
					return domain.Session{}, fmt.Errorf("commit idempotent chat session end: %w", commitErr)
				}
				return session, nil
			}
		}
		return domain.Session{}, transitionError("end chat session", err)
	}
	var userID *int64
	var enteredAt time.Time
	const visitQuery = `UPDATE visits SET left_at = $2, updated_at = $2 WHERE id = $1 AND left_at IS NULL
	RETURNING user_id, entered_at`
	if err := tx.QueryRow(ctx, visitQuery, session.VisitID, now.UTC()).Scan(&userID, &enteredAt); err != nil {
		return domain.Session{}, transitionError("finish visit", err)
	}
	if userID != nil {
		const rankQuery = `UPDATE registered_users SET chat_seconds = chat_seconds + GREATEST(EXTRACT(EPOCH FROM ($2::timestamp - $1::timestamp))::integer, 0) WHERE id = $3`
		if _, err := tx.Exec(ctx, rankQuery, enteredAt, now.UTC(), *userID); err != nil {
			return domain.Session{}, fmt.Errorf("increment chat time: %w", err)
		}
	}
	const messageQuery = `INSERT INTO room_messages (room_id, kind, author, body, theme_id, appearance, reactions, font_id, font_style, sent_at, inserted_at, updated_at)
	VALUES ($1, 'system', 'system', $2, 'vertigo', '{}'::jsonb, '{}'::jsonb, 'theme', 'normal', $3, $3, $3)`
	if _, err := tx.Exec(ctx, messageQuery, session.RoomID, "из чата выходит "+session.Nickname, now.UTC()); err != nil {
		return domain.Session{}, fmt.Errorf("persist departure: %w", err)
	}
	if err := tx.Commit(ctx); err != nil {
		return domain.Session{}, fmt.Errorf("commit chat session end: %w", err)
	}
	return session, nil
}

func (s Store) transition(ctx context.Context, action string, query string, args ...any) (domain.Session, error) {
	var session domain.Session
	if err := s.pool.QueryRow(ctx, query, args...).Scan(
		&session.ID, &session.RoomID, &session.IdentityKey, &session.Nickname, &session.Status, &session.Generation, &session.VisitID, &session.ReconnectDeadline,
	); err != nil {
		if errors.Is(err, pgx.ErrNoRows) {
			return domain.Session{}, ErrInvalidSession
		}
		return domain.Session{}, transitionError(action, err)
	}
	return session, nil
}

func hashSecret(secret string) string {
	sum := sha256.Sum256([]byte(secret))
	return hex.EncodeToString(sum[:])
}

func startError(action string, err error) error {
	return transitionError(action, err)
}

func transitionError(action string, err error) error {
	if errors.Is(err, pgx.ErrNoRows) {
		return ErrInvalidSession
	}
	return fmt.Errorf("%s: %w", action, err)
}
